### PR TITLE
Update Beam examples archetypes

### PR DESCRIPTION
**Pull request by peihe**
Includes: additinal examples code cleanups as a separate commit.
Prefer not to squash.

Verified examples and starter archetypes with:
  mvn archetype:generate \
      -DarchetypeArtifactId=beam-sdks-java-maven-archetypes-examples \
      -DarchetypeGroupId=org.apache.beam \
      -DarchetypeVersion=0.3.0-incubating-SNAPSHOT \
      -DgroupId=com.example \
      -DartifactId=first-beam \
      -Dversion="0.3.0-incubating-SNAPSHOT" \
      -DinteractiveMode=false \
      -Dpackage=org.apache.beam.examples

and did "mvn clean install" in first-beam project.

### CODE REVIEW

**Comment by peihe** [comment]
R: @tgroh

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/MinimalWordCount.java [hunk 1]
```diff
@@ -48,26 +49,34 @@
...
+    // CHANGE 1/3: Select a Beam runner, such as DataflowRunner or FlinkRunner.
+    // CHANGE 2/3: Specify runner-required options.
+    // For DataflowRunner, set project and temp location as follows:
+    //   DataflowPipelineOptions dataflowOptions = options.as(DataflowPipelineOptions.class);
+    //   dataflowOptions.setRunner(DataflowRunner.class);
+    //   dataflowOptions.setProject("SET_YOUR_PROJECT_ID_HERE");
+    //   dataflowOptions.setTempLocation("gs://SET_YOUR_BUCKET_NAME_HERE/AND_TEMP_DIRECTORY");
+    // For FlinkRunner, set the runner as follows. See {@code FlinkPipelineOptions}
+    // for more details.
+    //   options.as(FlinkPipelineOptions.class)
```
Comment (tgroh) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/MinimalWordCount.java]: just options.setRunner(FlinkRunner.class)

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleOptions.java [hunk 2]
```diff
@@ -17,16 +17,21 @@
...
-  @Description("Whether to keep jobs running on the Dataflow service after local process exit")
+public interface ExampleOptions extends PipelineOptions {
+  @Description("Whether to keep jobs running after local process exit")
   @Default.Boolean(false)
   boolean getKeepJobsRunning();
   void setKeepJobsRunning(boolean keepJobsRunning);
+
+  @Description("Number of workers to use when executing the injector pipeline")
+  @Default.Integer(1)
+  int getInjectorNumWorkers();
```
Comment (tgroh) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleOptions.java]: This appears to be unused.

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleUtils.java [hunk 3]
```diff
@@ -27,61 +30,46 @@
...
+import org.joda.time.Duration;
 
 /**
- * The utility class that sets up and tears down external resources, starts the Google Cloud Pub/Sub
- * injector, and cancels the streaming and the injector pipelines once the program terminates.
+ * The utility class that sets up and tears down external resources,
+ * and cancels the streaming pipelines once the program terminates.
  *
- * <p>It is used to run Dataflow examples, such as TrafficMaxLaneFlow and TrafficRoutes.
+ * <p>It is used to run Beam examples, such as TrafficMaxLaneFlow and TrafficRoutes.
```
Comment (tgroh) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleUtils.java]: These examples don't exist in the archetype. Should the references be updated to refer to `WindowedWordCount`?

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleUtils.java [hunk 4]
```diff
@@ -91,19 +79,26 @@ public DataflowExampleUtils(DataflowPipelineOptions options, boolean isUnbounded
...
+        FluentBackoff.DEFAULT
+            .withMaxRetries(3).withInitialBackoff(Duration.millis(200)).backoff();
+    Throwable lastException = null;
+    try {
+      do {
+        try {
+          setupPubsub();
+          setupBigQueryTable();
+          return;
+        } catch (GoogleJsonResponseException e) {
```
Comment (tgroh) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleUtils.java]: Is this an `IOException`? If so, catch it as such, as this is very GCP specific.

Comment (tgroh) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleUtils.java]: Actually, it is setting up GCP service connectors, so nevermind.

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleUtils.java [hunk 5]
```diff
@@ -91,19 +79,26 @@ public DataflowExampleUtils(DataflowPipelineOptions options, boolean isUnbounded
...
+          lastException = e;
+        }
+      } while (BackOffUtils.next(sleeper, backOff));
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      // Ignore InterruptedException
```
Comment (tgroh) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleUtils.java]: What happens if an `InterruptedException` is thrown before any other exception? The loop will terminate, so we'll throw a `RuntimeException`, so we should probably handle that case explicitly here.

Comment (peihe) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleUtils.java]: I think that is fine.
RuntimeException means the setup() fails. And, it keeps the thread interrupted. So, any code that care about the interruption status can check.

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/MinimalWordCount.java (hunk 1 above)
Comment (peihe) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/MinimalWordCount.java]: done

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleOptions.java (hunk 2 above)
Comment (peihe) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleOptions.java]: removed

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleUtils.java (hunk 3 above)
Comment (peihe) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleUtils.java]: removed

**Comment by peihe** [comment]
R: @dhalperi

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/MinimalWordCount.java [hunk 6]
```diff
@@ -48,26 +49,33 @@
...
+    // CHANGE 1/3: Select a Beam runner, such as DataflowRunner or FlinkRunner.
+    // CHANGE 2/3: Specify runner-required options.
+    // For DataflowRunner, set project and temp location as follows:
+    //   DataflowPipelineOptions dataflowOptions = options.as(DataflowPipelineOptions.class);
+    //   dataflowOptions.setRunner(DataflowRunner.class);
+    //   dataflowOptions.setProject("SET_YOUR_PROJECT_ID_HERE");
+    //   dataflowOptions.setTempLocation("gs://SET_YOUR_BUCKET_NAME_HERE/AND_TEMP_DIRECTORY");
+    // For FlinkRunner, set the runner as follows. See {@code FlinkPipelineOptions}
+    // for more details.
+    //   options.setRunner(FlinkRunner.class);
```
Comment (lukecwik) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/MinimalWordCount.java]: Differs from non archetype version:
    // for more details.
    //   options.as(FlinkPipelineOptions.class)
    //      .setRunner(FlinkRunner.class);

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/DebuggingWordCount.java [hunk 7]
```diff
@@ -28,16 +28,17 @@
 import org.apache.beam.sdk.options.PipelineOptionsFactory;
```
Comment (lukecwik) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/DebuggingWordCount.java]: Differs from non archetypes version:
Redundant import ${package}.WordCount, please remove

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleUtils.java [hunk 8]
```diff
@@ -27,61 +30,46 @@
...
 import org.apache.beam.sdk.util.Transport;
+import org.joda.time.Duration;
 
 /**
- * The utility class that sets up and tears down external resources, starts the Google Cloud Pub/Sub
- * injector, and cancels the streaming and the injector pipelines once the program terminates.
+ * The utility class that sets up and tears down external resources,
+ * and cancels the streaming pipelines once the program terminates.
  *
- * <p>It is used to run Dataflow examples, such as TrafficMaxLaneFlow and TrafficRoutes.
```
Comment (lukecwik) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleUtils.java]: Please update examples/java/src/main/java/org/apache/beam/examples/common/ExampleUtils.java to not mention TraffixMaxLaneFlow and TrafficRoutes so the archetype and non archetype versions are effectively the same

**Comment by lukecwik** [comment]
Minor comments and then LGTM

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/MinimalWordCount.java (hunk 6 above)
Comment (peihe) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/MinimalWordCount.java]: done

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/DebuggingWordCount.java (hunk 7 above)
Comment (peihe) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/DebuggingWordCount.java]: done

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleUtils.java (hunk 8 above)
Comment (peihe) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/main/java/common/ExampleUtils.java]: done

**Comment by peihe** [comment]
Fixed and verified the archetypes.
PTAL

Thanks Luke for helping me to debug

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/pom.xml [hunk 9]
```diff
@@ -200,12 +202,6 @@
...
 
-     <dependency>
-      <groupId>javax.servlet</groupId>
-      <artifactId>javax.servlet-api</artifactId>
-      <version>3.1.0</version>
-    </dependency>
-
     <!-- Add slf4j API frontend binding with JUL backend -->
     <dependency>
       <groupId>org.slf4j</groupId>
```
Comment (lukecwik) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/pom.xml]: update slf4j-api and slf4j-jdk14 version to 1.7.14 to match beam parent pom

File: sdks/java/maven-archetypes/starter/src/main/resources/archetype-resources/pom.xml [hunk 10]
```diff
@@ -55,7 +55,7 @@
     <dependency>
       <groupId>org.apache.beam</groupId>
       <artifactId>beam-sdks-java-core</artifactId>
-      <version>[0-incubating, 1-incubating)</version>
+      <version>0.3.0-incubating-SNAPSHOT</version>
     </dependency>
 
     <!-- slf4j API frontend binding with JUL backend -->
```
Comment (lukecwik) [inline sdks/java/maven-archetypes/starter/src/main/resources/archetype-resources/pom.xml]: update slf4j-api and slf4j-jdk14 version to 1.7.14 to match beam parent pom

File: sdks/java/maven-archetypes/starter/src/test/resources/projects/basic/reference/pom.xml [hunk 11]
```diff
@@ -55,7 +55,7 @@
     <dependency>
       <groupId>org.apache.beam</groupId>
       <artifactId>beam-sdks-java-core</artifactId>
-      <version>[0-incubating, 1-incubating)</version>
+      <version>0.3.0-incubating-SNAPSHOT</version>
     </dependency>
 
     <!-- slf4j API frontend binding with JUL backend -->
```
Comment (lukecwik) [inline sdks/java/maven-archetypes/starter/src/test/resources/projects/basic/reference/pom.xml]: update slf4j-api and slf4j-jdk14 version to 1.7.14 to match beam parent pom

File: sdks/java/maven-archetypes/examples/pom.xml [hunk 12]
```diff
@@ -66,4 +66,32 @@
       </plugins>
     </pluginManagement>
   </build>
+  <dependencies>
+    <!-- Adds a dependency on a specific version of the Beam SDK. -->
```
Comment (lukecwik) [inline sdks/java/maven-archetypes/examples/pom.xml]: You could drop the comments, I believe the package names explain their purpose

**Comment by lukecwik** [comment]
As a follow up, sync with Jason Kuster about making sure that Jenkins does the archetypes generate for starter and examples, and then executes mvn verify on those generated projects.

Similarly for travis.yml, make sure that the archetypes are generated for the starter and examples, and then executes mvn verify on those generated projects.

**Comment by lukecwik** [comment]
Thanks, merging

File: sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/pom.xml (hunk 9 above)
Comment (peihe) [inline sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/pom.xml]: done

File: sdks/java/maven-archetypes/starter/src/main/resources/archetype-resources/pom.xml (hunk 10 above)
Comment (peihe) [inline sdks/java/maven-archetypes/starter/src/main/resources/archetype-resources/pom.xml]: done

File: sdks/java/maven-archetypes/starter/src/test/resources/projects/basic/reference/pom.xml (hunk 11 above)
Comment (peihe) [inline sdks/java/maven-archetypes/starter/src/test/resources/projects/basic/reference/pom.xml]: done

File: sdks/java/maven-archetypes/examples/pom.xml (hunk 12 above)
Comment (peihe) [inline sdks/java/maven-archetypes/examples/pom.xml]: done

**Pull request closed without merging**